src/function/window/window_segment_tree.cpp

```diff
@@ -558,12 +570,19 @@ void WindowSegmentTreePart::EvaluateLeaves(const WindowSegmentTreeGlobalState &t
 	// current row
 	const bool add_curr_row = compute_left && frame_part == FramePart::RIGHT && exclude_mode == WindowExcludeMode::TIES;
 
+	// We need the full range of the frame to clamp
+	auto frame_begins = frame_part == FramePart::RIGHT ? bounds : begins;
+	auto frame_ends = frame_part == FramePart::LEFT ? bounds : ends;
+
 	for (idx_t rid = 0, cur_row = row_idx; rid < count; ++rid, ++cur_row) {
 		auto state_ptr = fdata[rid];
 
-		const auto begin = begin_on_curr_row ? cur_row + 1 : begins[rid];
-		const auto end = end_on_curr_row ? cur_row : ends[rid];
-		if (add_curr_row) {
+		const auto frame_begin = frame_begins[rid];
+		auto begin = MaxValue(begin_on_curr_row ? cur_row + 1 : begins[rid], frame_begin);
+
+		const auto frame_end = frame_ends[rid];
+		auto end = MinValue(end_on_curr_row ? cur_row : ends[rid], frame_end);
```

Comment:
These lines:
```c++
auto frame_begins = frame_part == FramePart::RIGHT ? bounds : begins;
auto frame_ends = frame_part == FramePart::LEFT ? bounds : ends;

auto begin = MaxValue(begin_on_curr_row ? cur_row + 1 : begins[rid], frame_begin);
auto end = MinValue(end_on_curr_row ? cur_row : ends[rid], frame_end);
```
Seem to be duplicated across `EvaluateUpperLevels` and `EvaluateLeaves`. Would it make sense to create functions for them? E.g., `GetFrameBeginsAndEnds`, `GetFrameBegin`, and `GetFrameEnd`